### PR TITLE
Use Spring Retry For Failed Requests Using Feign

### PR DESCRIPTION
**Pull request by ryanjbaxter**
Part of our continued effort to unify retry logic in Spring Cloud around Spring Retry.  Ties back to #1290.

### CODE REVIEW

**Comment by ryanjbaxter** [comment]
This is not likely ready for prime time but I wanted to get some eyes on these changes.

**Comment by ryanjbaxter** [comment]
spring-cloud/spring-cloud-sleuth#446 is dependent on this change

**Comment by ryanjbaxter** [comment]
One last comment ;)

These changes will also need to be merged into the 1.3.x branch.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/FeignClientsConfiguration.java [hunk 1]
```diff
@@ -16,6 +16,13 @@
 
 package org.springframework.cloud.netflix.feign;
 
+import com.netflix.hystrix.HystrixCommand;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/FeignClientsConfiguration.java]: Seems odd. Why a bunch of new imports?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java [hunk 2]
```diff
@@ -74,26 +95,32 @@ public RibbonResponse execute(RibbonRequest request, IClientConfig configOverrid
 		else {
 			options = new Request.Options(this.connectTimeout, this.readTimeout);
 		}
-		Response response = request.client().execute(request.toRequest(), options);
-		return new RibbonResponse(request.getUri(), response);
+		LoadBalancedRetryPolicy retryPolicy = loadBalancedRetryPolicyFactory.create(this.getClientName(), this);
+		retryTemplate.setRetryPolicy(retryPolicy == null ? new NeverRetryPolicy()
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java]: Seems like this should only be done once.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/ZuulProperties.java [hunk 3]
```diff
@@ -71,7 +68,7 @@
 	 * Flag for whether retry is supported by default (assuming the routes themselves
 	 * support it).
 	 */
-	private Boolean retryable;
+	private Boolean retryable = false;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/ZuulProperties.java]: seems unrelated.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/CachingSpringLoadBalancerFactory.java [hunk 4]
```diff
@@ -34,11 +35,16 @@
 public class CachingSpringLoadBalancerFactory {
 
 	private final SpringClientFactory factory;
+	private final RetryTemplate retryTemplate;
+	private final LoadBalancedRetryPolicyFactory loadBalancedRetryPolicyFactory;
 
 	private volatile Map<String, FeignLoadBalancer> cache = new ConcurrentReferenceHashMap<>();
 
-	public CachingSpringLoadBalancerFactory(SpringClientFactory factory) {
+	public CachingSpringLoadBalancerFactory(SpringClientFactory factory, RetryTemplate retryTemplate,
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/CachingSpringLoadBalancerFactory.java]: We should keep the old signature for backwards compatibility.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/apache/RibbonLoadBalancingHttpClient.java [hunk 5]
```diff
@@ -16,8 +16,12 @@
 
 package org.springframework.cloud.netflix.ribbon.apache;
 
-import java.net.URI;
-
+import com.netflix.client.RequestSpecificRetryHandler;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/apache/RibbonLoadBalancingHttpClient.java]: You need to fix your ordering in a number of files ;-)

**Comment by ryanjbaxter** [comment]
@spencergibb can you take another look when you get a chance?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/FeignClientsConfiguration.java [hunk 6]
```diff
@@ -107,7 +106,7 @@ public FormattingConversionService feignConversionService() {
 	@Scope("prototype")
 	@ConditionalOnMissingBean
 	public Feign.Builder feignBuilder() {
-		return Feign.builder();
+		return Feign.builder().retryer(Retryer.NEVER_RETRY);
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/FeignClientsConfiguration.java]: This should go in a `@Bean` definition.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java (hunk 2 above)
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java]: I'm still confused about where the request is used.

Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java]: So you are saying that the policy should only be created once?  The problem is we need the `Request` when `FeignRetryPolicy.open` is called so we can set the `ServiceInstance` in the Retry Context.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java]: I see, the retry policy created below uses the request. Could there be a problem since the retry policy is scoped to the request and the template is a field?

Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java]: Just trying to make sure I understand the concern....

If multiple requests are executed simultaneously you are concerned about switching out the retry policy mid request?

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java]: That's my worry

Comment (ryanjbaxter) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/ribbon/FeignLoadBalancer.java]: I agree, I guess the only good solution for now is to create a new `RetryTemplate` for each request.  I don't see any real downside to that, but I do think it can be handled better by Spring Retry.  I posed this in issue spring-projects/spring-retry#71.
